### PR TITLE
fix(dop): Handling some special cases and bugs of R&D workflow

### PR DESCRIPTION
**Pull request by littlejiancc**
#### What this PR does / why we need it:
Handling some special cases and bugs of R&D workflow

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJpdGVyYXRpb25JRHMiOlsxMjgwLDEyOTddLCJhc3NpZ25lZSI6WyIxMDAxMjYxIl19&id=319990&iterationID=1280&pId=0&type=BUG)
- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJpdGVyYXRpb25JRHMiOlsxMjgwLDEyOTddLCJhc3NpZ25lZSI6WyIxMDAxMjYxIl19&id=319305&iterationID=1280&pId=0&type=BUG)
- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJpdGVyYXRpb25JRHMiOlsxMjgwLDEyOTddLCJhc3NpZ25lZSI6WyIxMDAxMjYxIl19&id=319111&iterationID=1280&pId=0&type=BUG)

#### Specified Reviewers:

/assign @sfwn

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |      Handling some special cases and bugs of R&D workflow       |
| 🇨🇳 中文    |       处理研发工作流的一些特殊情况以及缺陷      |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: internal/apps/devflow/flow/flow.service.go [hunk 1]
```diff
@@ -229,47 +224,36 @@ func (s *Service) OperationMerge(ctx context.Context, req *pb.OperationMergeRequ
 	return &pb.OperationMergeResponse{}, nil
 }
 
-func (s *Service) GetTempBranchFromDevFlowRules(ctx context.Context, targetBranch string, projectID uint64) (string, error) {
-	result, err := s.p.DevFlowRule.GetDevFlowRulesByProjectID(ctx, &rulepb.GetDevFlowRuleRequest{
-		ProjectID: projectID,
+func (s *Service) getTempBranchFromFlowRule(ctx context.Context, projectID uint64, mrInfo *apistructs.MergeRequestInfo) (string, error) {
+	flowRule, err := s.p.DevFlowRule.GetFlowByRule(ctx, devflowrule.GetFlowByRuleRequest{
+		ProjectID:    projectID,
+		FlowType:     "multi_branch",
```
Comment (sfwn) [inline internal/apps/devflow/flow/flow.service.go:230]: define a type for FlowType

File: internal/apps/devflow/flow/flow.service.go [hunk 2]
```diff
@@ -582,37 +554,51 @@ func (s *Service) GetDevFlowInfo(ctx context.Context, req *pb.GetDevFlowInfoRequ
...
 				if err != nil {
 					return err
 				}
+				var baseCommit *apistructs.Commit
+				if tempBranch != "" {
+					if err := s.IdempotentCreateBranch(ctx, repoPath, v.TargetBranch, tempBranch); err != nil {
+						return err
+					}
+					if v.IsJoinTempBranch {
+						baseCommit, err = s.p.bdl.GetMergeBase(apis.GetUserID(ctx), apistructs.GittarMergeBaseRequest{
```
Comment (sfwn) [inline internal/apps/devflow/flow/flow.service.go:568]: do you need batch get merge base? loop may be very big.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: internal/apps/devflow/flow/flow.service.go (hunk 1 above)
Comment (littlejiancc) [reply internal/apps/devflow/flow/flow.service.go:230]: done

File: internal/apps/devflow/flow/flow.service.go (hunk 2 above)
Comment (littlejiancc) [reply internal/apps/devflow/flow/flow.service.go:568]: Generally, there is only one application

**Review by sfwn — APPROVED** [review]

**Pull request merged**